Use PSR Interfaces instead of guzzle

Zeichen32 opened the pull request:
This PR replace the fixed guzzle implementation with psr interfaces, so the developer can choose the used http client.

We need to update the docs or should  install a default client. Currently the user need to install the following packages to use guzzle.

```
"guzzlehttp/psr7": "^1.6",
"guzzlehttp/guzzle": "~6.0",
"php-http/guzzle6-adapter": "^2.0",
"http-interop/http-factory-guzzle": "^1.0"
```
By default this PR used an auto discovery to find possible http clients. To define the used http client manually the user need to provide an PSR HttpClient, an PSR Request Factory and a PSR Stream Factory as optional constructor parameters.

Closes #11

Zeichen32 commented:
@curquiza  Can you help me with this issue? The build failed because of a code style Issue. But unfortunately theire is no information about the line of the code style error.

```
Run vendor/bin/php-cs-fixer fix -v --config=.php_cs.dist --using-cache=no --dry-run
  shell: /bin/bash -e {0}
Loaded config default from ".php_cs.dist".
F.....................
Legend: ?-unknown, I-invalid file syntax (file ignored), S-skipped (cached or empty file), .-no changes, F-fixed, E-error
   1) meilisearch-php/src/HTTPRequest.php (blank_line_before_statement)

Checked all files in 0.315 seconds, 12.000 MB memory used
##[error]Process completed with exit code 8.
```

ppshobi commented:
Hello, Are you still facing issue on this?

Zeichen32 commented:
@ppshobi Unfortunately yes.

Zeichen32 commented:
@ppshobi @curquiza Code style error fixed and all tests are green. I think it is rdy to merge

curquiza commented:
Hello @Zeichen32! If your PR is indeed ready to be reviewed, could you remove the `WIP`? 🙂

Zeichen32 commented:
@curquiza I changed the title.

curquiza commented:
> @curquiza Can you help me with this issue? The build failed because of a code style Issue. But unfortunately theire is no information about the line of the code style error.

So sorryyyyyyy @Zeichen32 !!! I've just seen your message! Don't know why I haven't seen it before! My bad!
For future issues like this one, [join us to our Slack](https://slack.meilisearch.com/) and talk directly to me (Clémentine).

About the PR: thanks a lot for your work! It will be reviewed and merged soon 🙂

Again, sorry for the message I haven't seen. And thanks for your time on this PR! 😁

curquiza commented on composer.json:
```diff
@@ -13,7 +13,16 @@
     "require": {
         "php": "^7.2",
-        "guzzlehttp/guzzle": "^6.5"
+        "ext-json": "*",
+        "guzzlehttp/guzzle": "^6.5",
+        "php-http/client-implementation": "^1.0",
+        "psr/http-message": "^1.0",
+        "php-http/httplug": "^1.0|^2.0",
+        "php-http/message-factory": "^1.0",
+        "php-http/discovery": "^1.0",
+        "guzzlehttp/psr7": "^1.6",
+        "php-http/guzzle6-adapter": "^2.0",
+        "http-interop/http-factory-guzzle": "^1.0"
```
For me, it works with only these dependencies:
```json
"require": {
        "php": "^7.2",
        "ext-json": "*",
        "php-http/discovery": "^1.7",
        "php-http/guzzle6-adapter": "^2.0",
        "http-interop/http-factory-guzzle": "^1.0"
    },
```

Zeichen32 commented:
@curquiza 

> For me, it works with only these dependencies:  

You are right there! I updated the dependencies.

> Could you add a small part in the README

Do you mean as an example or as a new main point in the README? For now i have added it as an example how to replace the HTTP client.

ppshobi commented on src/HTTPRequest.php:
```diff
@@ -2,36 +2,68 @@
+     */
+    private $streamFactory;
+
+    public function __construct($url, $api_key = null, ClientInterface $httpClient = null, RequestFactoryInterface $requestFactory = null, StreamFactoryInterface $streamFactory = null)
```
This not a constructive criticism but I see that in a lot of places the variable names are snake_case and you have used a mix of snake_cased and camelCased variable names. Maybe we could stick to one practice? I suggest camelCase to be used everywhere.

Zeichen32 commented:
@ppshobi I'm unable to squash all commits in one. I'm not 100% sure why. Maybe i should create a new PR with a single commit?

curquiza commented:
@Zeichen32, no problem if you didn't succeed to squash, I will do it for you, I have a "squash and merge" option 😉

curquiza commented on README.md:
````diff
@@ -251,6 +252,27 @@ $index->search('prince', ['limit' => 2, 'filters' => "title = 'Le Petit Prince'
 $index->search('hobbit', ['limit' => 2, 'filters' => 'title = "The Hitchhiker\'s Guide to the Galaxy" OR author = "J. R. R. Tolkien"']);
 ```
 
+### Replace HTTP client
````
```suggestion
## Replace HTTP client

```

curquiza commented on README.md:
```diff
@@ -33,6 +33,7 @@
   - [Documents](#documents)
   - [Update status](#update-status)
   - [Search](#search)
+  - [Replace HTTP client](#replace-http-client)
```
```suggestion
- [Replace HTTP client](#replace-http-client)
```

curquiza commented on README.md:
````diff
@@ -251,6 +252,27 @@ $index->search('prince', ['limit' => 2, 'filters' => "title = 'Le Petit Prince'
 ```
 
+### Replace HTTP client
+By default [Guzzle](http://docs.guzzlephp.org/en/6.5/) is used for all HTTP request.
````
```suggestion
By default [Guzzle](http://docs.guzzlephp.org/en/6.5/) is used for all HTTP requests.<br>
```

curquiza replied:
(The markdown will not create a newl ine, you have to specify it)

curquiza commented on README.md:
```diff
@@ -251,6 +252,27 @@ $index->search('prince', ['limit' => 2, 'filters' => "title = 'Le Petit Prince'
 
+### Replace HTTP client
+By default [Guzzle](http://docs.guzzlephp.org/en/6.5/) is used for all HTTP request.
+But you can replace the used HTTP client with any [PSR-18](https://www.php-fig.org/psr/psr-18/) compatible client.
```
```suggestion
But you can replace the used HTTP client with any [PSR-18](https://www.php-fig.org/psr/psr-18/) compatible client.<br>
```

curquiza commented on README.md:
```diff
@@ -251,6 +252,27 @@ $index->search('prince', ['limit' => 2, 'filters' => "title = 'Le Petit Prince'
+A list with compatible HTTP clients and client adapters can be found at [php-http.org](http://docs.php-http.org/en/latest/clients.html).
+
+For example if you want to use the [Symfony HttpClient](https://symfony.com/doc/current/components/http_client.html)
+use the following code:
```
```suggestion
```

curquiza commented on README.md:
```diff
@@ -251,6 +252,27 @@ $index->search('prince', ['limit' => 2, 'filters' => "title = 'Le Petit Prince'
+But you can replace the used HTTP client with any [PSR-18](https://www.php-fig.org/psr/psr-18/) compatible client.
+A list with compatible HTTP clients and client adapters can be found at [php-http.org](http://docs.php-http.org/en/latest/clients.html).
+
+For example if you want to use the [Symfony HttpClient](https://symfony.com/doc/current/components/http_client.html)
```
```suggestion
For example, if you want to use the [Symfony HttpClient](https://symfony.com/doc/current/components/http_client.html) use the following code:
```

Zeichen32 commented:
@curquiza 

> Is it guzzle?

Yes, because we require the Guzzle6 Adapter as dependeny in the composer.json. So if no other clients are installed it only find Guzzle.

I've update the readme with your suggestions.
